File: electron/backend/process.js
/**
 * Backend Process Management
 * Handles Python executable detection, environment setup, and process spawning
 */
const { spawn, execSync } = require('child_process');
const path = require('path');
const fs = require('fs');
const { app } = require('electron');
const { getPackagesDir } = require('./config');

/**
 * Get the Python executable path based on platform
 * @returns {string} Path to Python executable
 */
function getPythonPath() {
  const isPackaged = app.isPackaged;

  if (process.platform === 'win32') {
    if (isPackaged) {
      // Bundled Python in resources folder
      return path.join(process.resourcesPath, 'python-embed', 'python.exe');
    } else {
      // Development: check for local python-embed folder
      const localEmbed = path.join(__dirname, '..', '..', 'python-embed', 'python.exe');
      if (fs.existsSync(localEmbed)) {
        return localEmbed;
      }
      // Development: check for standard .venv folder
      const dotVenv = path.join(__dirname, '..', '..', '.venv', 'Scripts', 'python.exe');
      if (fs.existsSync(dotVenv)) {
        return dotVenv;
      }
      // Fall back to system Python
      return 'python';
    }
  } else {
    // Linux/macOS
    if (isPackaged) {
      // Check for bundled venv in resources folder
      const bundledPython = path.join(process.resourcesPath, 'python-venv', 'bin', 'python');
      if (fs.existsSync(bundledPython)) {
        return bundledPython;
      }
    } else {
      // Development: check for local python-venv-linux folder
      const localVenv = path.join(__dirname, '..', '..', 'python-venv-linux', 'bin', 'python');
      if (fs.existsSync(localVenv)) {
        return localVenv;
      }
      // Development: check for standard .venv folder
      const dotVenv = path.join(__dirname, '..', '..', '.venv', 'bin', 'python');
      if (fs.existsSync(dotVenv)) {
        return dotVenv;
      }
    }
    // Fall back to system Python
    return 'python';
  }
}

/**
 * Get the working directory for the Python process
 * @returns {string} Working directory path
 */
function getWorkingDirectory() {
  if (app.isPackaged) {
    // Check if we have bundled resources (api folder in resources/)
    const bundledApiPath = path.join(process.resourcesPath, 'api');
    if (fs.existsSync(bundledApiPath)) {
      // Windows or Linux with bundled venv: api folder is in resources/
      return process.resourcesPath;
    }
    // Fallback: api folder is in resources/app (unbundled Linux/macOS)
    return path.join(process.resourcesPath, 'app');
  }
  return path.join(__dirname, '..', '..');
}

/**
 * Get environment variables for Python subprocess
 * @param {string|null} portableDataDir - Portable data directory or null
 * @returns {Object} Environment variables object
 */
function getPythonEnv(portableDataDir) {
  const baseEnv = { ...process.env, PYTHONUNBUFFERED: '1' };
  const packagesDir = getPackagesDir();

  // Ensure packages directory exists
  if (!fs.existsSync(packagesDir)) {
    fs.mkdirSync(packagesDir, { recursive: true });
  }

  // Add portable data directory if in portable mode
  if (portableDataDir) {
    baseEnv.LOCALBOORU_PORTABLE_DATA = portableDataDir;
  }

  if (process.platform === 'win32') {
    const pythonPath = getPythonPath();
    const pythonDir = path.dirname(pythonPath);

    if (app.isPackaged || fs.existsSync(path.join(pythonDir, 'python.exe'))) {
      // Using bundled Python
      // Add onnxruntime capi folder to PATH for DLL loading
      const onnxCapi = path.join(packagesDir, 'onnxruntime', 'capi');
      const bundledOnnxCapi = path.join(pythonDir, 'Lib', 'site-packages', 'onnxruntime', 'capi');

      // Bundled video pipeline tools (ffmpeg, vapoursynth)
      const rootDir = app.isPackaged ? process.resourcesPath : path.join(__dirname, '..', '..');
      const ffmpegDir = path.join(rootDir, 'ffmpeg');
      const vsDir = path.join(rootDir, 'vapoursynth');
      const vsPluginsDir = path.join(vsDir, 'vs-plugins');

      // Build PATH with video tools prepended (if present)
      let pathPrefix = `${onnxCapi};${bundledOnnxCapi};${pythonDir};${path.join(pythonDir, 'Scripts')}`;
      if (fs.existsSync(ffmpegDir)) {
        pathPrefix = `${ffmpegDir};${pathPrefix}`;
      }
      if (fs.existsSync(vsDir)) {
        pathPrefix = `${vsDir};${pathPrefix}`;
      }

      const env = {
        ...baseEnv,
        PATH: `${pathPrefix};${process.env.PATH}`,
        PYTHONHOME: pythonDir,
        PYTHONPATH: `${packagesDir};${getWorkingDirectory()};${path.join(pythonDir, 'Lib', 'site-packages')}`,
        LOCALBOORU_PACKAGED: '1',
        LOCALBOORU_PACKAGES_DIR: packagesDir
      };

      // Set bundled video tool env vars if directories exist
      if (fs.existsSync(path.join(vsDir, 'python.exe'))) {
        env.LOCALBOORU_VS_PYTHON = path.join(vsDir, 'python.exe');
      }
      if (fs.existsSync(vsPluginsDir)) {
        env.LOCALBOORU_SVP_PLUGIN_PATH = vsPluginsDir;
        env.LOCALBOORU_VS_PLUGIN_PATH = vsPluginsDir;
      }

      return env;
    }
  } else {
    // Linux/macOS
    const pythonPath = getPythonPath();
    const pythonDir = path.dirname(path.dirname(pythonPath)); // go up from bin/python to venv root

    // Check if using bundled venv
    const isBundledVenv = pythonPath.includes('python-venv');

    if (isBundledVenv) {
      // Using bundled venv - find the site-packages directory
      const libDir = path.join(pythonDir, 'lib');
      let sitePackagesPath = '';

      // Find the python3.X directory inside lib
      if (fs.existsSync(libDir)) {
        const entries = fs.readdirSync(libDir);
        for (const entry of entries) {
          if (entry.startsWith('python3.')) {
            const spPath = path.join(libDir, entry, 'site-packages');
            if (fs.existsSync(spPath)) {
              sitePackagesPath = spPath;
              break;
            }
          }
        }
      }

      return {
        ...baseEnv,
        // Add bundled venv's bin to PATH
        PATH: `${path.join(pythonDir, 'bin')}:${process.env.PATH}`,
        // Include persistent packages, working directory, and bundled site-packages
        PYTHONPATH: `${packagesDir}:${getWorkingDirectory()}${sitePackagesPath ? ':' + sitePackagesPath : ''}`,
        LOCALBOORU_PACKAGED: app.isPackaged ? '1' : '',
        LOCALBOORU_PACKAGES_DIR: packagesDir
      };
    } else {
      // System Python with pyenv support
      const homeDir = process.env.HOME || process.env.USERPROFILE;
      const pyenvPath = `${homeDir}/.pyenv/shims:${homeDir}/.pyenv/bin`;
      return {
        ...baseEnv,
        PATH: `${pyenvPath}:${process.env.PATH}`,
        PYTHONPATH: `${packagesDir}:${getWorkingDirectory()}`,
        LOCALBOORU_PACKAGED: app.isPackaged ? '1' : '',
        LOCALBOORU_PACKAGES_DIR: packagesDir
      };
    }
  }

  return baseEnv;
}

/**
 * Spawn the uvicorn backend process
 * @param {Object} options - Spawn options
 * @param {string} options.bindHost - Host to bind to
 * @param {number} options.port - Port to bind to
 * @param {string|null} options.portableDataDir - Portable data directory or null
 * @returns {ChildProcess} Spawned process
 */
function spawnBackend({ bindHost, port, portableDataDir }) {
  const pythonPath = getPythonPath();
  const cwd = getWorkingDirectory();
  const env = getPythonEnv(portableDataDir);

  console.log('[Backend] Python path:', pythonPath);
  console.log('[Backend] Working directory:', cwd);
  console.log('[Backend] Binding to:', bindHost);

  // Spawn uvicorn via python -m for better compatibility
  return spawn(pythonPath, [
    '-m', 'uvicorn',
    'api.main:app',
    '--host', bindHost,
    '--port', String(port)
  ], {
    cwd: cwd,
    stdio: ['ignore', 'pipe', 'pipe'],
    env: env,
    // Windows-specific: hide console window
    windowsHide: true
  });
}

/**
 * Kill uvicorn processes aggressively
 * Used for cleanup on startup and shutdown
 */
function killUvicornProcesses() {
  try {
    if (process.platform === 'win32') {
      execSync('taskkill /F /IM python.exe /FI "WINDOWTITLE eq uvicorn*" 2>nul', { stdio: 'ignore', timeout: 5000 });
    } else {
      // Kill any uvicorn process for api.main:app specifically
      execSync('pkill -9 -f "uvicorn api.main:app" 2>/dev/null || true', { stdio: 'ignore', timeout: 5000 });
    }
  } catch (e) {
    // Ignore errors - process might not exist
  }
}

/**
 * Force kill a specific process by PID
 * @param {number} pid - Process ID to kill
 */
function forceKillProcess(pid) {
  try {
    if (process.platform === 'win32') {
      execSync(`taskkill /pid ${pid} /T /F`, { stdio: 'ignore' });
    } else {
      process.kill(pid, 'SIGKILL');
    }
  } catch (e) {
    // Process may already be dead
  }
}

/**
 * Gracefully terminate a process (Windows)
 * @param {number} pid - Process ID to terminate
 */
function gracefulKillProcess(pid) {
  try {
    if (process.platform === 'win32') {
      // First try graceful tree kill (no /F flag)
      execSync(`taskkill /pid ${pid} /T`, { stdio: 'ignore', timeout: 3000 });
    }
  } catch (e) {
    // If that fails, the caller should handle force kill
  }
}

module.exports = {
  getPythonPath,
  getWorkingDirectory,
  getPythonEnv,
  spawnBackend,
  killUvicornProcesses,
  forceKillProcess,
  gracefulKillProcess
};

File: src-tauri/src/backend/process.rs
//! Backend Process Management
//!
//! Handles Python executable detection, environment setup, and process spawning.
//! Mirrors electron/backend/process.js

use std::env;
use std::fs;
use std::path::PathBuf;
use std::process::{Command, Child, Stdio};
use std::collections::HashMap;

use super::config::get_packages_dir;

/// Get the Python executable path based on platform
pub fn get_python_path(is_packaged: bool) -> String {
    if cfg!(target_os = "windows") {
        if is_packaged {
            // Bundled Python in resources folder
            // For Tauri, this would be in the resource directory
            if let Ok(exe_path) = env::current_exe() {
                if let Some(parent) = exe_path.parent() {
                    let bundled = parent.join("python-embed").join("python.exe");
                    if bundled.exists() {
                        return bundled.to_string_lossy().to_string();
                    }
                }
            }
        }

        // Development: check for local venv
        let dev_paths = [
            PathBuf::from("python-embed/python.exe"),
            PathBuf::from(".venv/Scripts/python.exe"),
        ];

        for path in &dev_paths {
            if path.exists() {
                return path.to_string_lossy().to_string();
            }
        }

        // Fall back to system Python
        "python".to_string()
    } else {
        // Linux/macOS
        if is_packaged {
            // Check for bundled venv in resources folder
            if let Ok(exe_path) = env::current_exe() {
                if let Some(parent) = exe_path.parent() {
                    let bundled = parent.join("python-venv").join("bin").join("python");
                    if bundled.exists() {
                        return bundled.to_string_lossy().to_string();
                    }
                }
            }
        }

        // Development: check for local venv
        let dev_paths = [
            PathBuf::from("python-venv-linux/bin/python"),
            PathBuf::from(".venv/bin/python"),
        ];

        for path in &dev_paths {
            if path.exists() {
                return path.to_string_lossy().to_string();
            }
        }

        // Fall back to system Python
        "python".to_string()
    }
}

/// Get the working directory for the Python process
pub fn get_working_directory(is_packaged: bool) -> PathBuf {
    if is_packaged {
        // For packaged app, look for api folder relative to executable
        if let Ok(exe_path) = env::current_exe() {
            if let Some(parent) = exe_path.parent() {
                let api_path = parent.join("api");
                if api_path.exists() {
                    return parent.to_path_buf();
                }
                // Fallback: check resources subdirectory
                let resources_api = parent.join("resources").join("api");
                if resources_api.exists() {
                    return parent.join("resources");
                }
            }
        }
    }

    // Development: use current directory or find project root
    if let Ok(cwd) = env::current_dir() {
        // Check if api folder exists in cwd
        if cwd.join("api").exists() {
            return cwd;
        }
        // Check parent directories
        let mut current = cwd.clone();
        for _ in 0..3 {
            if let Some(parent) = current.parent() {
                if parent.join("api").exists() {
                    return parent.to_path_buf();
                }
                current = parent.to_path_buf();
            }
        }
        return cwd;
    }

    PathBuf::from(".")
}

/// Build environment variables for the Python subprocess
pub fn get_python_env(
    is_packaged: bool,
    portable_data_dir: Option<&PathBuf>,
) -> HashMap<String, String> {
    let mut env_vars: HashMap<String, String> = env::vars().collect();

    // Always unbuffer Python output for real-time logging
    env_vars.insert("PYTHONUNBUFFERED".to_string(), "1".to_string());

    let packages_dir = get_packages_dir();

    // Ensure packages directory exists
    if !packages_dir.exists() {
        let _ = fs::create_dir_all(&packages_dir);
    }

    // Add portable data directory if in portable mode
    if let Some(data_dir) = portable_data_dir {
        env_vars.insert(
            "LOCALBOORU_PORTABLE_DATA".to_string(),
            data_dir.to_string_lossy().to_string(),
        );
    }

    let working_dir = get_working_directory(is_packaged);
    let python_path = get_python_path(is_packaged);

    if cfg!(target_os = "windows") {
        let python_dir = PathBuf::from(&python_path)
            .parent()
            .map(|p| p.to_path_buf())
            .unwrap_or_default();

        if is_packaged || python_dir.join("python.exe").exists() {
            // Using bundled Python
            let site_packages = python_dir.join("Lib").join("site-packages");
            let onnx_capi = packages_dir.join("onnxruntime").join("capi");
            let bundled_onnx_capi = site_packages.join("onnxruntime").join("capi");

            // Build PATH
            let current_path = env_vars.get("PATH").cloned().unwrap_or_default();
            let new_path = format!(
                "{};{};{};{};{}",
                onnx_capi.to_string_lossy(),
                bundled_onnx_capi.to_string_lossy(),
                python_dir.to_string_lossy(),
                python_dir.join("Scripts").to_string_lossy(),
                current_path
            );
            env_vars.insert("PATH".to_string(), new_path);
            env_vars.insert("PYTHONHOME".to_string(), python_dir.to_string_lossy().to_string());

            let python_path_val = format!(
                "{};{};{}",
                packages_dir.to_string_lossy(),
                working_dir.to_string_lossy(),
                site_packages.to_string_lossy()
            );
            env_vars.insert("PYTHONPATH".to_string(), python_path_val);
            env_vars.insert("LOCALBOORU_PACKAGED".to_string(), "1".to_string());
            env_vars.insert("LOCALBOORU_PACKAGES_DIR".to_string(), packages_dir.to_string_lossy().to_string());
        }
    } else {
        // Linux/macOS
        let python_path_buf = PathBuf::from(&python_path);
        let is_bundled_venv = python_path.contains("python-venv");

        if is_bundled_venv {
            // Using bundled venv
            let venv_root = python_path_buf
                .parent() // bin
                .and_then(|p| p.parent()) // venv root
                .unwrap_or(&python_path_buf);

            // Find site-packages
            let lib_dir = venv_root.join("lib");
            let mut site_packages_path = String::new();

            if lib_dir.exists() {
                if let Ok(entries) = fs::read_dir(&lib_dir) {
                    for entry in entries.flatten() {
                        let name = entry.file_name();
                        let name_str = name.to_string_lossy();
                        if name_str.starts_with("python3.") {
                            let sp_path = entry.path().join("site-packages");
                            if sp_path.exists() {
                                site_packages_path = sp_path.to_string_lossy().to_string();
                                break;
                            }
                        }
                    }
                }
            }

            // Update PATH
            let current_path = env_vars.get("PATH").cloned().unwrap_or_default();
            let bin_path = venv_root.join("bin");
            env_vars.insert("PATH".to_string(), format!("{}:{}", bin_path.to_string_lossy(), current_path));

            // Set PYTHONPATH
            let mut python_path_val = format!(
                "{}:{}",
                packages_dir.to_string_lossy(),
                working_dir.to_string_lossy()
            );
            if !site_packages_path.is_empty() {
                python_path_val = format!("{}:{}", python_path_val, site_packages_path);
            }
            env_vars.insert("PYTHONPATH".to_string(), python_path_val);
        } else {
            // System Python with pyenv support
            let home_dir = env::var("HOME").unwrap_or_default();
            let pyenv_path = format!("{}/.pyenv/shims:{}/.pyenv/bin", home_dir, home_dir);
            let current_path = env_vars.get("PATH").cloned().unwrap_or_default();
            env_vars.insert("PATH".to_string(), format!("{}:{}", pyenv_path, current_path));

            let python_path_val = format!(
                "{}:{}",
                packages_dir.to_string_lossy(),
                working_dir.to_string_lossy()
            );
            env_vars.insert("PYTHONPATH".to_string(), python_path_val);
        }

        env_vars.insert(
            "LOCALBOORU_PACKAGED".to_string(),
            if is_packaged { "1" } else { "" }.to_string(),
        );
        env_vars.insert(
            "LOCALBOORU_PACKAGES_DIR".to_string(),
            packages_dir.to_string_lossy().to_string(),
        );
    }

    env_vars
}

/// Spawn the uvicorn backend process
pub fn spawn_backend(
    bind_host: &str,
    port: u16,
    is_packaged: bool,
    portable_data_dir: Option<&PathBuf>,
) -> std::io::Result<Child> {
    let python_path = get_python_path(is_packaged);
    let cwd = get_working_directory(is_packaged);
    let env_vars = get_python_env(is_packaged, portable_data_dir);

    log::info!("[Backend] Python path: {}", python_path);
    log::info!("[Backend] Working directory: {:?}", cwd);
    log::info!("[Backend] Binding to: {}", bind_host);

    // Spawn uvicorn via python -m for better compatibility
    let mut cmd = Command::new(&python_path);
    cmd.args([
        "-m", "uvicorn",
        "api.main:app",
        "--host", bind_host,
        "--port", &port.to_string(),
    ])
    .current_dir(&cwd)
    .envs(env_vars)
    .stdin(Stdio::null())
    .stdout(Stdio::piped())
    .stderr(Stdio::piped());

    // Windows-specific: hide console window
    #[cfg(target_os = "windows")]
    {
        use std::os::windows::process::CommandExt;
        cmd.creation_flags(0x08000000); // CREATE_NO_WINDOW
    }

    cmd.spawn()
}

/// Kill uvicorn processes aggressively
/// Used for cleanup on startup and shutdown
#[cfg(not(target_os = "windows"))]
pub fn kill_uvicorn_processes() {
    // Kill any uvicorn process for api.main:app specifically
    let _ = Command::new("pkill")
        .args(["-9", "-f", "uvicorn api.main:app"])
        .output();
}

#[cfg(target_os = "windows")]
pub fn kill_uvicorn_processes() {
    // Windows: try to kill python processes running uvicorn
    let _ = Command::new("taskkill")
        .args(["/F", "/IM", "python.exe", "/FI", "WINDOWTITLE eq uvicorn*"])
        .output();
}

/// Force kill a specific process by PID
#[cfg(not(target_os = "windows"))]
pub fn force_kill_process(pid: u32) {
    let _ = Command::new("kill")
        .args(["-9", &pid.to_string()])
        .output();
}

#[cfg(target_os = "windows")]
pub fn force_kill_process(pid: u32) {
    let _ = Command::new("taskkill")
        .args(["/pid", &pid.to_string(), "/T", "/F"])
        .output();
}

/// Gracefully terminate a process
#[cfg(not(target_os = "windows"))]
pub fn graceful_kill_process(pid: u32) {
    // Send SIGINT first (Ctrl+C equivalent)
    let _ = Command::new("kill")
        .args(["-2", &pid.to_string()])
        .output();
}

#[cfg(target_os = "windows")]
pub fn graceful_kill_process(pid: u32) {
    // First try graceful tree kill (no /F flag)
    let _ = Command::new("taskkill")
        .args(["/pid", &pid.to_string(), "/T"])
        .output();
}
